### ISSUE TITLE
super verbose error messages unrelated to PyF

### ISSUE THREAD

**Issue by ghorn** [guibou/PyF#18]
This is an example where `PyF` works perfectly fine, but GHC gives a warning that exposes the internals of `PyF`:
 ```haskell
{-# LANGUAGE QuasiQuotes #-}

module Kitty.KittyProto.Types ( Foo(..) ) where

import qualified Data.Text as T
import Kitty.PyF ( fmt )

data Foo =
   Foo
   { fieldA :: ()
   , fieldB :: T.Text
   }

yolo :: Foo
yolo =
   Foo
   { fieldB = [fmt|hello what's up {x}|]
   }
   where
     x :: T.Text
     x = T.pack "hi"
```
Ghci helpfully reminds me to initialize `fieldA`:
```
ErrorMessageExplosion.hs:15:1: error: [-Wunused-top-binds, -Werror=unused-top-binds]
    Defined but not used: ‘yolo’
   |
15 | yolo =
   | ^^^^

ErrorMessageExplosion.hs:16:3: error: [-Wmissing-fields, -Werror=missing-fields]
    • Fields of ‘Foo’ not initialised: fieldA
    • In the expression:
        Foo
          {fieldB = (PyF.magicFormat
                       (Data.Text.Internal.Builder.fromString "hello what's up "
                          <>
                            Data.Text.Internal.Builder.fromString
                              ((\ v_PyF
                                  -> case
                                         (PyF.Internal.QQ.categorise
                                            (Data.Proxy.Proxy ::
                                               Data.Proxy.Proxy PyF.Internal.QQ.EnableForString))
                                           v_PyF
                                     of
                                       PyF.Internal.QQ.Integral i_PyF
                                         -> ((((PyF.Internal.QQ.formatAnyIntegral
                                                  PyF.Formatters.Decimal)
                                                 PyF.Formatters.Minus)
                                                (PyF.Internal.QQ.newPadding
                                                   PyF.Internal.PythonSyntax.PaddingDefault))
                                               ((PyF.Internal.QQ.toGrp Nothing) 3))
                                              i_PyF
                                       PyF.Internal.QQ.Fractional f_PyF
                                         -> (((((PyF.Internal.QQ.formatAnyFractional
                                                   PyF.Formatters.Generic)
                                                  PyF.Formatters.Minus)
                                                 (PyF.Internal.QQ.newPadding
                                                    PyF.Internal.PythonSyntax.PaddingDefault))
                                                ((PyF.Internal.QQ.toGrp Nothing) 3))
                                               (PyF.Internal.QQ.changePrec'
                                                  PyF.Internal.PythonSyntax.PrecisionDefault))
                                              f_PyF
                                       PyF.Internal.QQ.StringType f_PyF
                                         -> ((PyF.Formatters.formatString
                                                (PyF.Internal.QQ.newPaddingForString
                                                   PyF.Internal.PythonSyntax.PaddingDefault))
                                               (PyF.Internal.QQ.changePrec'
                                                  PyF.Internal.PythonSyntax.PrecisionDefault))
                                              f_PyF)
                                 x)))}
      In an equation for ‘yolo’:
          yolo
            = Foo
                {fieldB = (PyF.magicFormat
                             (Data.Text.Internal.Builder.fromString "hello what's up "
                                <>
                                  Data.Text.Internal.Builder.fromString
                                    ((\ v_PyF
                                        -> case
                                               (PyF.Internal.QQ.categorise
                                                  (Data.Proxy.Proxy ::
                                                     Data.Proxy.Proxy PyF.Internal.QQ.EnableForString))
                                                 v_PyF
                                           of
                                             PyF.Internal.QQ.Integral i_PyF
                                               -> ((((PyF.Internal.QQ.formatAnyIntegral
                                                        PyF.Formatters.Decimal)
                                                       PyF.Formatters.Minus)
                                                      (PyF.Internal.QQ.newPadding
                                                         PyF.Internal.PythonSyntax.PaddingDefault))
                                                     ((PyF.Internal.QQ.toGrp Nothing) 3))
                                                    i_PyF
                                             PyF.Internal.QQ.Fractional f_PyF
                                               -> (((((PyF.Internal.QQ.formatAnyFractional
                                                         PyF.Formatters.Generic)
                                                        PyF.Formatters.Minus)
                                                       (PyF.Internal.QQ.newPadding
                                                          PyF.Internal.PythonSyntax.PaddingDefault))
                                                      ((PyF.Internal.QQ.toGrp Nothing) 3))
                                                     (PyF.Internal.QQ.changePrec'
                                                        PyF.Internal.PythonSyntax.PrecisionDefault))
                                                    f_PyF
                                             PyF.Internal.QQ.StringType f_PyF
                                               -> ((PyF.Formatters.formatString
                                                      (PyF.Internal.QQ.newPaddingForString
                                                         PyF.Internal.PythonSyntax.PaddingDefault))
                                                     (PyF.Internal.QQ.changePrec'
                                                        PyF.Internal.PythonSyntax.PrecisionDefault))
                                                    f_PyF)
                                       x)))}
            where
                x :: T.Text
                x = T.pack "hi"
   |
16 |   Foo
   |   ^^^...
```
Unfortunately it also gives me some context, including the expansion of the `PyF` quasiquoter.

Is this a fundamental limitation of quasiquoters, or is there something that could be done? It's pretty funny that if the inputs to the quasiquoter are badly formed, I get a beautiful error message from `PyF`, but if the inputs are well formed and another part of the program is bad, the error messages is unreadible.

**Comment by guibou** [comment]
Thank you for reporting this.

This is AFAIK a limitation of the quasi quoter and we'll see the generated code in any circumstance. However I can work on minimizing the size of the generated code to something more readable.

**Comment by guibou** [comment]
@ghorn 

I've just merged #30 which contains a lot of improvements on the following issue.

In your example, the splice was:

```
(PyF.magicFormat
                       (Data.Text.Internal.Builder.fromString "hello what's up "
                          <>
                            Data.Text.Internal.Builder.fromString
                              ((\ v_PyF
                                  -> case
                                         (PyF.Internal.QQ.categorise
                                            (Data.Proxy.Proxy ::
                                               Data.Proxy.Proxy PyF.Internal.QQ.EnableForString))
                                           v_PyF
                                     of
                                       PyF.Internal.QQ.Integral i_PyF
                                         -> ((((PyF.Internal.QQ.formatAnyIntegral
                                                  PyF.Formatters.Decimal)
                                                 PyF.Formatters.Minus)
                                                (PyF.Internal.QQ.newPadding
                                                   PyF.Internal.PythonSyntax.PaddingDefault))
                                               ((PyF.Internal.QQ.toGrp Nothing) 3))
                                              i_PyF
                                       PyF.Internal.QQ.Fractional f_PyF
                                         -> (((((PyF.Internal.QQ.formatAnyFractional
                                                   PyF.Formatters.Generic)
                                                  PyF.Formatters.Minus)
                                                 (PyF.Internal.QQ.newPadding
                                                    PyF.Internal.PythonSyntax.PaddingDefault))
                                                ((PyF.Internal.QQ.toGrp Nothing) 3))
                                               (PyF.Internal.QQ.changePrec'
                                                  PyF.Internal.PythonSyntax.PrecisionDefault))
                                              f_PyF
                                       PyF.Internal.QQ.StringType f_PyF
                                         -> ((PyF.Formatters.formatString
                                                (PyF.Internal.QQ.newPaddingForString
                                                   PyF.Internal.PythonSyntax.PaddingDefault))
                                               (PyF.Internal.QQ.changePrec'
                                                  PyF.Internal.PythonSyntax.PrecisionDefault))
                                              f_PyF)
                                 x)))
```

And is now:

```
    magicFormat
      (Builder.fromString "hello what's up "
         <>
           Builder.fromString
             (((((PyF.Internal.QQ.formatAny PyF.Formatters.Minus)
                   PyF.Internal.QQ.PaddingDefaultK)
                  Nothing)
                 Nothing)
                x))
```

The duplication of the dump of the slice in the error message won't be removed, that's GHC way of dumping errors. However we moved from 37 lines of cryptic code to 9 line of roughly understable code.

More improvement can be done on that topic and I'll address them later. I've opened issue #31 to track it and I close your issue considering that your problem is mostly solved.
